File: packages/sdk-python/openpassport/envelope.py
"""Create signed message envelopes."""
from __future__ import annotations

import secrets
from datetime import datetime, timezone
from typing import Any

from openpassport.canonical import canonicalize
from openpassport.keys import base64url_encode, sign_bytes


def create_envelope(
    from_id: str,
    passport_url: str,
    body: Any,
    private_key: bytes,
    to_id: str | None = None,
) -> dict[str, Any]:
    """Create a signed message envelope."""
    nonce = secrets.token_hex(16)
    now = datetime.now(timezone.utc)

    unsigned: dict[str, Any] = {
        "openpassport_message": "0.1.0",
        "from": from_id,
        "passportUrl": passport_url,
        "timestamp": now.strftime("%Y-%m-%dT%H:%M:%SZ"),
        "nonce": nonce,
        "body": body,
    }

    if to_id is not None:
        unsigned["to"] = to_id

    canonical = canonicalize(unsigned)
    data = canonical.encode("utf-8")
    sig = sign_bytes(data, private_key)

    return {**unsigned, "signature": base64url_encode(sig)}

File: packages/sdk-js/src/envelope.ts
/**
 * Create and sign OpenPassport message envelopes.
 *
 * Every agent-to-agent message is wrapped in a signed envelope
 * that includes a timestamp and nonce for replay protection.
 */
import { canonicalize, type Envelope } from "@openpassport/spec";
import { base64UrlEncode, sign } from "./keys.js";

export interface CreateEnvelopeOptions {
  from: string;
  to?: string;
  passportUrl: string;
  body: unknown;
}

export async function createEnvelope(
  options: CreateEnvelopeOptions,
  privateKey: Uint8Array,
): Promise<Envelope> {
  // 128-bit random hex nonce for replay protection
  const nonce = Array.from(crypto.getRandomValues(new Uint8Array(16)))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");

  const unsigned: Omit<Envelope, "signature"> = {
    openpassport_message: "0.1.0",
    from: options.from,
    ...(options.to ? { to: options.to } : {}),
    passportUrl: options.passportUrl,
    timestamp: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
    nonce,
    body: options.body,
  };

  const canonical = canonicalize(unsigned as Record<string, unknown>);
  const bytes = new TextEncoder().encode(canonical);
  const sig = await sign(bytes, privateKey);

  return {
    ...unsigned,
    signature: base64UrlEncode(sig),
  };
}
